### PR TITLE
feat(builders): browser-sync for ssr-dev-server

### PR DESCRIPTION
**Pull request by manfredsteyer**
After building both, the server bundles and the client bundles,
the ssr-dev-server starts the node server as well as a
browser-sync server. After the bundles changed, the node
server is restarted and the browser-sync server notifies
the browser.

### CODE REVIEW

File: modules/builders/BUILD.bazel [hunk 1]
```diff
@@ -31,7 +31,10 @@ ts_library(
     deps = [
         "@npm//@angular-devkit/architect",
         "@npm//@angular-devkit/core",
+        "@npm//@types/browser-sync",
+        "@npm//browser-sync",
         "@npm//rxjs",
+        "@npm//tree-kill",
```
Comment (alan-agius4) [inline modules/builders/BUILD.bazel]: Can we not use tree-kill please? This is a bit redundant in this case

File: modules/builders/src/ssr-dev-server/utils.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            if (err) {
+                obs.error(err);
+                return;
+            }
+
+            obs.next({ stdout, stderr });
+            obs.complete();
+        });
+
+        return () => treeKill(proc.pid, 'SIGTERM');
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/utils.ts]: Instead of tree-kill use

```
if (!process.killed){
 process.kill();
}

File: modules/builders/src/ssr-dev-server/index.ts [hunk 3]
```diff
@@ -45,11 +52,59 @@ export function execute(
...
+      map(_ => ([browserOutput, serverOutput]))
+    );
+}
+
+function startBrowserSync(port: number): browserSync.BrowserSyncInstance {
+
+  return browserSync.init({
+    proxy: `localhost:${port}`,
+    // TODO: Make this port configurable
+    port: 4200,
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: Can we do this? It’s just a matter of adding a couple of lines

File: bazel-universal-fork [hunk 4]
```diff
@@ -0,0 +1 @@
+/home/manfred/.cache/bazel/_bazel_manfred/3c5b42f541f2eebfc8b037dd0cc1c8d0/execroot/nguniversal
```
Comment (alan-agius4) [inline bazel-universal-fork]: Revert this file pls

File: package.json [hunk 5]
```diff
@@ -57,12 +57,14 @@
...
     "@hapi/hapi": "^18.4.0",
     "@schematics/angular": "^9.0.0-next.14",
+    "@types/browser-sync": "^2.26.1",
     "@types/express": "^4.17.1",
     "@types/fs-extra": "^8.0.0",
     "@types/hapi__hapi": "^18.2.5",
     "@types/hapi__inert": "^5.2.0",
     "@types/jasmine": "^3.4.4",
     "@types/node": "^12.11.1",
+    "browser-sync": "^2.26.7",
```
Comment (alan-agius4) [inline package.json]: This should be added as a dependency of the builders package.

File: modules/builders/src/ssr-dev-server/index.ts [hunk 6]
```diff
@@ -45,11 +52,59 @@ export function execute(
...
+      });
+    }),
   );
 }
 
+function startNodeServer(browserOutput: BuilderOutput, serverOutput: BuilderOutput, port: number) {
+  const path = serverOutput.outputPath;
+  const env = { ...process.env, PORT: '' + port };
+
+  return execAsObservable(`node ${path}/main.js`, { env })
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: The path here needs to be escaped due to special characters.

If I recall correctly this path is absolute. Can you confirm?

File: modules/builders/src/ssr-dev-server/index.ts [hunk 7]
```diff
@@ -45,11 +52,59 @@ export function execute(
...
+    catchError(error => {
+      return of({
+        success: false,
+        error: error
+      });
+    }),
   );
 }
 
+function startNodeServer(browserOutput: BuilderOutput, serverOutput: BuilderOutput, port: number) {
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: BrowserOutput is redundant here.

File: modules/builders/src/ssr-dev-server/index.ts [hunk 8]
```diff
@@ -45,11 +52,59 @@ export function execute(
...
+  const path = serverOutput.outputPath;
+  const env = { ...process.env, PORT: '' + port };
+
+  return execAsObservable(`node ${path}/main.js`, { env })
+    .pipe(
+      startWith({ stdout: '', stderr: '' }),
+        // Emit a signal after the process has been started
+      map(_ => ([browserOutput, serverOutput]))
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: Why not map in the main function.

In here we also need to handle server errors, if the server emits and error we shouldn’t crash.

Also probably stdout should be pipped to avoid having log messages in each restart such as `node server is listening on port 3333`

File: modules/builders/src/ssr-dev-server/index.ts [hunk 9]
```diff
@@ -45,11 +52,59 @@ export function execute(
...
+    );
+}
+
+function startBrowserSync(port: number): browserSync.BrowserSyncInstance {
+
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: We need to do some logs like the port the server is listing too, etc. Similar to `ng serve`.

Will get you the message later.

File: modules/builders/src/ssr-dev-server/index.ts [hunk 10]
```diff
@@ -45,11 +52,59 @@ export function execute(
...
+  return from(getAvailablePort()).pipe(
+    tap(port => {
+      nodeServerPort = port;
+    }),
+    switchMap(_ => merge(browserTargetRun, serverTargetRun, 1)),
+    bufferCount(2),
     switchMap(([b, s]) => zip(b.output, s.output)),
-    map(([b, s]) => ({ success: b.success && s.success }))
+    switchMap(([b, s]) => startNodeServer(b, s, nodeServerPort)),
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: If both builds weren’t successful we should try to spawn the server

File: modules/builders/src/ssr-dev-server/index.ts (hunk 9 above)
Comment (alan-agius4) [reply modules/builders/src/ssr-dev-server/index.ts]:
```js
            context.logger.info(tags.oneLine`
              **
              Angular Universal Live Development Server is listening on http://localhost:${port},
              open your browser on http://localhost:${port}
              **
            `);

```

File: modules/builders/src/ssr-dev-server/index.ts [hunk 11]
```diff
@@ -45,11 +52,59 @@ export function execute(
...
-    pairwise(),
+  let bsInstance: browserSync.BrowserSyncInstance;
+  let nodeServerPort: number;
+
+  return from(getAvailablePort()).pipe(
+    tap(port => {
+      nodeServerPort = port;
+    }),
+    switchMap(_ => merge(browserTargetRun, serverTargetRun, 1)),
+    bufferCount(2),
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: This is not fully correct, a user can do a change which triggers only a server builder or client build, example changing the `server.ts`.

File: modules/builders/src/ssr-dev-server/index.ts [hunk 12]
```diff
@@ -45,11 +52,59 @@ export function execute(
...
+    switchMap(([b, s]) => startNodeServer(b, s, nodeServerPort)),
+    tap(_ => {
+      if (!bsInstance) {
+        bsInstance = startBrowserSync(nodeServerPort);
+      } else {
+        bsInstance.reload();
+      }
+    }),
+    map(([b, s]) => ({ success: b.success && s.success })),
+    catchError(error => {
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: Nit 
```js
    catchError(error => {
      return of({
        success: false,
        error: error
      });
    }),
```

```js
    catchError(error => of({
        success: false,
        error,
     });
 ),

```

File: modules/builders/src/ssr-dev-server/index.ts [hunk 13]
```diff
@@ -45,11 +52,59 @@ export function execute(
     watch: true,
   }));
 
-  return merge(browserTargetRun, serverTargetRun, 1).pipe(
-    pairwise(),
+  let bsInstance: browserSync.BrowserSyncInstance;
+  let nodeServerPort: number;
+
+  return from(getAvailablePort()).pipe(
+    tap(port => {
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: Nit: 

```js
  tap(port => nodeServerPort = port),
```

File: modules/builders/src/ssr-dev-server/index.ts (hunk 3 above)
Comment (manfredsteyer) [reply modules/builders/src/ssr-dev-server/index.ts]: +1

File: modules/builders/src/ssr-dev-server/index.ts (hunk 6 above)
Comment (manfredsteyer) [reply modules/builders/src/ssr-dev-server/index.ts]: good point

Yes, it's absolute.

File: modules/builders/src/ssr-dev-server/index.ts (hunk 7 above)
Comment (manfredsteyer) [reply modules/builders/src/ssr-dev-server/index.ts]: What do you mean? (browserOuput vs serverOutput)

File: modules/builders/BUILD.bazel (hunk 1 above)
Comment (manfredsteyer) [reply modules/builders/BUILD.bazel]: You mean `tree-kill` instead of `@npm//tree-kill` ?

File: modules/builders/src/ssr-dev-server/index.ts (hunk 11 above)
Comment (manfredsteyer) [reply modules/builders/src/ssr-dev-server/index.ts]: IMHO it's ok in this case, b/c this very switchMap only runs once at the start. After each change, the other ones run.

Do I see this right?

However, how to deal with this special case within the other switchMaps? Using `zip` could be an issue in cases where only one build changes. Right?

File: modules/builders/src/ssr-dev-server/index.ts (hunk 8 above)
Comment (manfredsteyer) [reply modules/builders/src/ssr-dev-server/index.ts]: What do you mean?

File: modules/builders/BUILD.bazel (hunk 1 above)
Comment (alan-agius4) [reply modules/builders/BUILD.bazel]: No, remove tree-kill as a dependency all together

File: modules/builders/src/ssr-dev-server/index.ts (hunk 7 above)
Comment (alan-agius4) [reply modules/builders/src/ssr-dev-server/index.ts]: the browserOutput object here is redundant., it is used by the `map()` but really we don't need that to be done here.

Comment (alan-agius4) [reply modules/builders/src/ssr-dev-server/index.ts]: See comment below.

File: modules/builders/src/ssr-dev-server/index.ts (hunk 8 above)
Comment (alan-agius4) [reply modules/builders/src/ssr-dev-server/index.ts]: 1) If the server crashes, we should crash the watch, so we need to gracefully catch the error
2) if the server throw an error, we need to handle the stderr and print it in the console
3) I don't think that we should do this in these method: `map(_ => ([browserOutput, serverOutput]))` strictly speaking this function doesn't care about the browserOutput and providing such imports for mapping seems like an overkill and not the best in terms of maintainability to some extent. I think a better and cleaner approach is to use `pipe`.

```ts
function startNodeServer(outputPath: string, port: number): Observable<void> {
```


Some preudo code;

```
from(getAvailablePort()).pipe(
    tap(port => nodeServerPort = port),
    switchMap(_ => merge(browserTargetRun, serverTargetRun, 1)),
    bufferCount(2),
    switchMap(([b, s]) => zip(b.output, s.output)),
    map(([b, s]) => ({
        success: b.success && s.success,
        error: b.error || s.error,
        serverOutputPath: b.outputPath || '',
    }),
    switchMap(buildEvent => {
        if (!buildEvent.success) {
            return buildEvent;
        }

        return startNodeServer(buildEvent.serverOutputPath, nodeServerPort)
            .pipe(
                tap(() => {
                    if (!bsInstance) {
                        bsInstance = startBrowserSync(nodeServerPort);
                    } else {
                        bsInstance.reload();
                    }
                }),
                // If a server error occuires we should print it but we shouldn't crash the watch
                catchError(() => NEVER),
                mapTo(buildEvent)
            );
    }),
    catchError(error => {
        return of({
            success: false,
            error: error
        });
    }),
);

```

File: modules/builders/src/ssr-dev-server/index.ts [hunk 14]
```diff
@@ -45,11 +52,59 @@ export function execute(
...
+      nodeServerPort = port;
+    }),
+    switchMap(_ => merge(browserTargetRun, serverTargetRun, 1)),
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: Question: why `_` instead of `()` is this a personal preference?

File: modules/builders/src/ssr-dev-server/index.ts (hunk 11 above)
Comment (alan-agius4) [reply modules/builders/src/ssr-dev-server/index.ts]: You are absolutely right. I think to overcome this we need to do something a bit different in terms of the RXJS patterns used here which would also simplify the logic.

My idea would be to do something along;
```ts
 // the below async parts might make sense to be moved to an init function or something along that route
    const browserTarget = targetFromTargetString(options.browserTarget);
    const serverTarget = targetFromTargetString(options.serverTarget);
  
    const browserTargetRun = await context.scheduleTarget(browserTarget, {
      watch: true,
      serviceWorker: false,
    }));
  
    const serverTargetRun = await context.scheduleTarget(serverTarget, {
      watch: true,
    });
  
    const nodeServerPort = await getAvailablePort();

    const bsInstance = browserSync.init({
        // let's no force refreshes close to each other
        reloadDebounce: 250
    });

    const serverRunOutput =  serverRun.output.pipe(
        // check if success
        // if not don't restart
        tap(() => startNodeServer(b, s, nodeServerPort),
    );

    return combineLatest(
        serverRunOutput,
        browserRun.output
    )
    .pipe(
        // if both success otherwise don't
        tap(() => bsInstance.reload());
    );
```


NB;
- With the above we should also stop the jobs
```ts
await Promise.all([browserTargetRun.stop(), serverTargetRun.stop()]);
```


.

File: modules/builders/BUILD.bazel (hunk 1 above)
Comment (manfredsteyer) [reply modules/builders/BUILD.bazel]: If I remove it, I'll get this:

```
modules/builders/src/ssr-dev-server/utils.ts:11:27 - error TS2307: Cannot find module 'tree-kill'.

11 import * as treeKill from 'tree-kill';
                             ~~~~~~~~~~~
```

Comment (alan-agius4) [reply modules/builders/BUILD.bazel]: Yeah you need to refactor the code to not use tree-kill see, comments above.

Comment (alan-agius4) [reply modules/builders/BUILD.bazel]: https://github.com/angular/universal/pull/1330#discussion_r345536651

Comment (manfredsteyer) [reply modules/builders/BUILD.bazel]: Ah, now I see. Thx.

File: modules/builders/src/ssr-dev-server/index.ts (hunk 14 above)
Comment (manfredsteyer) [reply modules/builders/src/ssr-dev-server/index.ts]: No specific reason. I personally prefere ``_``. Shall I change it to ``()``?

File: modules/builders/src/ssr-dev-server/index.ts (hunk 11 above)
Comment (manfredsteyer) [reply modules/builders/src/ssr-dev-server/index.ts]: I've implemented this idea in a minimally invasive way: ``combineLatest`` instead of ``zip`` + ``debounceTime``.

This is also beneficial for the case where the user is saving several files in a row.

My experiments also show that starting/ reloading browserSync when the node server is not fully available leads to issues (at least under windows). This is also solved by a short ``debounceTime``.

File: modules/builders/src/ssr-dev-server/index.ts [hunk 15]
```diff
@@ -36,20 +46,126 @@ export function execute(
...
+        // Emit a signal after the process has been started
+    );
+}
+
+function startBrowserSync(
+  syncServerPort: number,
+  nodeServerPort: number,
+  logger: logging.LoggerApi): Observable<browserSync.BrowserSyncInstance> {
+
+  const syncServerPort$ = (syncServerPort === 0) ?
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: I would strongly recommend in not using rxjs for this stuff. It’s an overkill as really these is no reason why this method needs to be reactive.

Async/await will suffice.

File: modules/builders/src/ssr-dev-server/index.ts [hunk 16]
```diff
@@ -36,20 +46,126 @@ export function execute(
...
+    tap(([b, s]) => {
+      serverTargetRun = s;
+      browserTargetRun = b;
+    }),
+    switchMap(([b, s]) => combineLatest(b.output, s.output)),
+    switchMap(([b, s]) => {
+      if (!b.success || !s.success) {
+        return of([b, s]);
+      }
+      return startNodeServer(s, nodeServerPort).pipe(
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: Would this start the node server twice for every change? Ie: when the browser and server is changed.


Hence, that was the reason why in my example is left the streams separate.

File: modules/builders/src/ssr-dev-server/index.ts [hunk 17]
```diff
@@ -36,20 +46,126 @@ export function execute(
...
+            return startBrowserSync(options.port, nodeServerPort, context.logger).pipe(
+              tap(instance => bsInstance = instance),
+              mapTo([b, s])
+            );
+          } else {
+            bsInstance.reload();
+            return of([b, s]);
+          }
+        }),
+        catchError(err => {
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: The catch error should be for the nodeServer part since If that throws we don’t want to continue with browser sync etc...

File: modules/builders/src/ssr-dev-server/index.ts [hunk 18]
```diff
@@ -36,20 +46,126 @@ export function execute(
...
+          console.error('Error in browser sync', err);
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: Don’t use console logs, use the logger from context.

File: modules/builders/src/ssr-dev-server/index.ts [hunk 19]
```diff
@@ -36,20 +46,126 @@ export function execute(
...
+      if (!b.success || !s.success) {
+        return of([b, s]);
+      }
+      return startNodeServer(s, nodeServerPort).pipe(
+        tap(result => {
+          if (result.stderr) {
+            console.error('Error in node server', result.stderr);
+          }
+        }),
+        debounceTime(500),
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: This is rather a high debounce time, why is this needed?

File: modules/builders/src/ssr-dev-server/index.ts [hunk 20]
```diff
@@ -36,20 +46,126 @@ export function execute(
...
+            return of([b, s]);
+          }
+        }),
+        catchError(err => {
+          console.error('Error in browser sync', err);
+          return NEVER;
+        }),
+      );
+    }),
+    map(([b, s]) => ({ success: b.success && s.success })),
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: You are not handling errors in this map, see previous example: 
See: https://github.com/angular/universal/pull/1330#discussion_r346191990

```ts
    map(([b, s]) => ({
        success: b.success && s.success,
        error: b.error || s.error,
        serverOutputPath: b.outputPath || '',
    }),
```

File: modules/builders/src/ssr-dev-server/index.ts [hunk 21]
```diff
@@ -36,20 +46,126 @@ export function execute(
...
+    }),
+    switchMap(([b, s]) => combineLatest(b.output, s.output)),
+    switchMap(([b, s]) => {
+      if (!b.success || !s.success) {
+        return of([b, s]);
+      }
+      return startNodeServer(s, nodeServerPort).pipe(
+        tap(result => {
+          if (result.stderr) {
+            console.error('Error in node server', result.stderr);
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: Having an `stderr` not empty doesn't necessary mean there was error.
stderr, can be used for various reasons;

1) error
2) warnings
3) progress indicators

etc..

Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: don't use console.logs, use the logger in context

File: modules/builders/src/ssr-dev-server/index.ts (hunk 14 above)
Comment (alan-agius4) [reply modules/builders/src/ssr-dev-server/index.ts]: If you don't mind :) Since that's typically the common pattern used in our repos.

File: modules/builders/src/ssr-dev-server/index.ts (hunk 17 above)
Comment (alan-agius4) [reply modules/builders/src/ssr-dev-server/index.ts]: Sorry I just realised I did it wrong in here https://github.com/angular/universal/pull/1330#discussion_r346191990 but correct in here https://github.com/angular/universal/pull/1330#discussion_r346245761 since it's outside of rxjs

File: modules/builders/src/ssr-dev-server/index.ts (hunk 15 above)
Comment (manfredsteyer) [reply modules/builders/src/ssr-dev-server/index.ts]: +1 Will do it in a new PR.

File: modules/builders/src/ssr-dev-server/index.ts (hunk 11 above)
Comment (manfredsteyer) [reply modules/builders/src/ssr-dev-server/index.ts]: Now, we only restart the node server if the server bundle is rebuilt.

File: bazel-universal-fork (hunk 4 above)
Comment (manfredsteyer) [reply bazel-universal-fork]: it's reverted now.

File: modules/builders/src/ssr-dev-server/index.ts [hunk 22]
```diff
@@ -36,20 +45,132 @@ export function execute(
...
+    })),
+    finalize(() => {
+      serverTargetRun.stop();
+      browserTargetRun.stop();
+    })
   );
 }
 
+function startNodeServer(serverOutput: BuilderOutput, port: number) {
+  const path = normalize(serverOutput.outputPath + '/main.js');
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: `join` is better in this case.

File: modules/builders/src/ssr-dev-server/index.ts [hunk 23]
```diff
@@ -36,20 +45,132 @@ export function execute(
...
+    switchMap(([br, sr]) => {
+      const server$ = sr.output.pipe(
+        switchMap(s => {
+          if (s.error) {
+            return of(s);
+          }
+          return startNodeServer(s, nodeServerPort).pipe(
+            map(() => s),
+            catchError(err => {
+              context.logger.error('Error in browser sync', err);
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: The error message here doesn’t seem right as this is unrelated to browser sync.

Maybe better something like 
`A server error has occurred.`

File: modules/builders/src/ssr-dev-server/index.ts [hunk 24]
```diff
@@ -36,20 +45,132 @@ export function execute(
...
+    tap(port => nodeServerPort = port),
+    switchMap(() => combineLatest(browserTargetRun$, serverTargetRun$)),
+    tap(([br, sr]) => {
+      serverTargetRun = sr;
+      browserTargetRun = br;
+    }),
+    switchMap(([br, sr]) => {
+      const server$ = sr.output.pipe(
+        switchMap(s => {
+          if (s.error) {
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: You shouldn’t check `.error` check `.success` instead. We might have an unsuccessful build  without an error msg.

File: modules/builders/src/ssr-dev-server/index.ts [hunk 25]
```diff
@@ -36,20 +45,132 @@ export function execute(
...
+        return startBrowserSync(options.port, nodeServerPort, context.logger).pipe(
+          tap(instance => bsInstance = instance),
+          mapTo([b, s])
+        );
+      } else {
+        bsInstance.reload();
+        return of([b, s]);
+      }
+    }),
+    map(([b, s]) => ({
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: This should be at the very being not at the very end. So that you avoid adding doing code like 

`if (!b.success || !s.success)`

Or 

`of([b, s])`

File: modules/builders/src/ssr-dev-server/index.ts [hunk 26]
```diff
@@ -36,20 +45,132 @@ export function execute(
...
-    pairwise(),
-    switchMap(([b, s]) => zip(b.output, s.output)),
-    map(([b, s]) => ({ success: b.success && s.success }))
+  let bsInstance: browserSync.BrowserSyncInstance;
+  let nodeServerPort: number;
+  let serverTargetRun: BuilderRun;
+  let browserTargetRun: BuilderRun;
+
+  return from(getAvailablePort()).pipe(
+    tap(port => nodeServerPort = port),
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: Move this after the build, the more time you want to assign the port the greater the risk it will become unavailable.

File: bazel-universal-fork (hunk 4 above)
Comment (alan-agius4) [reply bazel-universal-fork]: Bad file!!! Keeps re-appearing!!!

File: modules/builders/src/ssr-dev-server/index.ts (hunk 22 above)
Comment (alan-agius4) [reply modules/builders/src/ssr-dev-server/index.ts]: Since you are on Windows can you check that it works on paths with spaces and special characters such as `(` without escaping the path?

File: modules/builders/src/ssr-dev-server/index.ts (hunk 25 above)
Comment (manfredsteyer) [reply modules/builders/src/ssr-dev-server/index.ts]: As it is currently a big stream, I think I need it at the end. Before, I need ``[b, s]`` in the individual steps.

Or do I miss sth here?

File: modules/builders/src/ssr-dev-server/index.ts [hunk 27]
```diff
@@ -36,20 +45,132 @@ export function execute(
   const browserTarget = targetFromTargetString(options.browserTarget);
   const serverTarget = targetFromTargetString(options.serverTarget);
 
-  const browserTargetRun = from(context.scheduleTarget(browserTarget, {
+  const browserTargetRun$ = from(context.scheduleTarget(browserTarget, {
     watch: true,
     serviceWorker: false,
   }));
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: Can you also add `extractCss: true` here please

//cc @vikerman

File: modules/builders/src/ssr-dev-server/index.ts (hunk 25 above)
Comment (alan-agius4) [reply modules/builders/src/ssr-dev-server/index.ts]: You don't need `[b, s]` in the steps :)

```ts
return from(getAvailablePort()).pipe(
    tap(port => nodeServerPort = port),
    switchMap(() => combineLatest(browserTargetRun$, serverTargetRun$)),
    tap(([br, sr]) => {
        serverTargetRun = sr;
        browserTargetRun = br;
    }),
    switchMap(([br, sr]) => {
        const server$ = sr.output.pipe(
            switchMap(s => {
                if (s.error) {
                    return of(s);
                }
                return startNodeServer(s, nodeServerPort).pipe(
                    map(() => s),
                    catchError(err => {
                        context.logger.error('Error in browser sync', err);
                        return NEVER;
                    }),
                );
            }));

        return combineLatest(br.output, server$);
    }),
    map(([b, s]) => ({
        success: b.success && s.success,
        error: b.error || s.error,
        serverOutputPath: b.outputPath || '',
    } as BuilderOutput)),
    switchMap(buildEvent => {
        if (!bsInstance && buildEvent.success) {
            return startBrowserSync(options.port, nodeServerPort, context.logger).pipe(
                tap(instance => bsInstance = instance),
                mapTo(buildEvent),
            );
        }

        if (buildEvent.success) {
            bsInstance.reload();
        }
        return of(buildEvent);
    }),
    catchError(error => of({
        success: false,
        error: mapErrorToMessage(error)
    })),
    finalize(() => {
        serverTargetRun.stop();
        browserTargetRun.stop();
    })
);
```

File: modules/builders/src/ssr-dev-server/index.ts [hunk 28]
```diff
@@ -36,20 +45,132 @@ export function execute(
...
+    map(([b, s]) => ({
+      success: b.success && s.success,
+      error: b.error || s.error,
+      serverOutputPath: b.outputPath || '',
+    } as BuilderOutput)),
+    catchError(error => of({
+        success: false,
+        error: mapErrorToMessage(error)
+    })),
+    finalize(() => {
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/index.ts]: Since you are using obs, I think you don't need to stop since it should be handled in the tear down logic. I'll check tmr.

File: modules/builders/src/ssr-dev-server/schema.json [hunk 29]
```diff
@@ -11,6 +11,11 @@
     "serverTarget": {
       "type": "string",
       "description": "Browser target to build."
+    },
+    "port": {
+      "type": "number",
+      "default": 4200,
+      "description": "Port to start the sync server at. Default is 4200. Pass 0 to get a dynamically assigned port"
```
Comment (alan-agius4) [inline modules/builders/src/ssr-dev-server/schema.json]: Nit: missing full stop at the end

File: bazel-universal-fork (hunk 4 above)
Comment (manfredsteyer) [reply bazel-universal-fork]: Should be gone now. If you click at the file name above, the link is broken.

Comment (alan-agius4) [reply bazel-universal-fork]: Still here and the link is not broken

**Comment by alan-agius4** [comment]
@googlebot I consent.

File: integration/express-engine-ivy/package.json [hunk 30]
```diff
@@ -10,11 +10,11 @@
     "postinstall": "webdriver-manager update --gecko false --standalone false $CHROMEDRIVER_VERSION_ARG",
     "ng": "ng",
     "start": "ng serve",
-    "e2e": "ng e2e",
-    "test": "npm run build && npm run serve:test",
-    "serve:test": "concurrently \"npm run serve:ssr\" \"npm run e2e\" --kill-others --success first",
+    "e2e": "sleep 60 && ng e2e",
```
[the remark this answers is not in the document]
Comment (alan-agius4) [reply integration/express-engine-ivy/package.json]: The sleep is needed because since now we build and run e2e concurrently e2e will run prior to the server being started.

While I did a method to probe for the app-root, I did the sleep to wait prior to even start this. https://github.com/angular/universal/pull/1330/files#diff-73654ffb2ae5aa618d303608b80b9775R15

File: integration/express-engine-ivy/package.json [hunk 31]
```diff
@@ -10,11 +10,11 @@
...
+    "test": "npm run serve:test",
+    "serve:test": "concurrently \"npm run serve:ssr:dev\" \"npm run e2e\" --kill-others --success first",
     "serve:ssr": "node dist/express-engine-ivy/server/main.js",
-    "serve:ssr-dev-server": "ng run express-engine-ivy:ssr-serve",
+    "serve:ssr:dev": "ng run express-engine-ivy:serve-ssr:production",
```
[the remark this answers is not in the document]
Comment (alan-agius4) [reply integration/express-engine-ivy/package.json]: That's just for the integration tests since we run e2e integration tests in production. we can rename the `script` name. 

Maybe for test purposes we can name it 
```
serve:ssr:prod
```
